crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -789,6 +799,18 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    fn infer_match_pattern_definition(
+        &mut self,
+        _pattern: &ast::Pattern,
+        _index: u32,
+        definition: Definition<'db>,
+    ) {
+        // TODO(dhruvmanila): The correct way to infer types here is to perform structural matching
+        // against the subject expression type (available in `self.index.expression_ty(subject)`)
```

Comment:
I think you mean `self.types.expression_ty(subject)` here (`self.index` is the semantic index, which doesn't have an `expression_ty` method.)

It's a bit misleading to say "available in ...", because it won't be, if we are inferring a single match pattern definition region, unless we ensure it gets put there. And we should only infer the type of the subject once, even if we have many match-pattern definitions referring to it. This is why we add the subject as a standalone expression in the semantic index; so that we can query its type directly using a cached Salsa query. So the approach we will want here is that we will call the Salsa query `infer_expression_types(db, expression)`, where we get `expression` (which is an `Expression` ingredient) from `self.index.expression(the_subject_ast_expr)`, and extend our own inference with the result. Once that's done then we can look up its type in `self.types.expression_ty(subject.scoped_ast_id())`. (This is all similar to what's currently done in `infer_assignment_definition`).

It's fine for all of this to remain a TODO in this PR! I would just re-word the comment a bit:
```suggestion
        // against the subject expression type (which we can query via `infer_expression_types`)
```

Reply:
That makes sense, thanks for the explanation. I'll also add a reference to `self.infer_assignment_definition`.